cmd/juju/charmhub/data.go

```diff
@@ -92,28 +99,50 @@ func convertCharm(in interface{}) (*Charm, error) {
 	}, nil
 }
 
-func convertChannels(in map[string]charmhub.Channel, series string) map[string]Channel {
+func convertChannels(in map[string]charmhub.Channel) map[string]Channel {
 	out := make(map[string]Channel, len(in))
 	for k, v := range in {
-		// If the platforms contains the all series, then we need to display
-		// all the channels for that given
-		allSeries := channelSeries(v.Platforms).Contains(SeriesAll)
-		if !allSeries && (series != "" && !channelSeries(v.Platforms).Contains(series)) {
-			continue
-		}
-
 		out[k] = Channel{
 			ReleasedAt: v.ReleasedAt,
 			Track:      v.Track,
 			Risk:       v.Risk,
 			Revision:   v.Revision,
 			Size:       v.Size,
 			Version:    v.Version,
+			Series:     channelSeries(v.Platforms).SortedValues(),
+			Arches:     channelArches(v.Platforms).SortedValues(),
 		}
 	}
+
 	return out
 }
 
+func filterChannels(in map[string]charmhub.Channel, architecture, series string) map[string]charmhub.Channel {
+	allArch := architecture == ArchAll
+	allSeries := series == SeriesAll
+
+	// If we're searching for everything then we can skip the filtering logic
+	// and return immediately.
+	if allArch && allSeries {
+		return in
+	}
+
+	// Channels that match any part of the criteria should be witnessed and
+	// kept.
+	witnessed := make(map[string]charmhub.Channel)
+
+	for k, v := range in {
+		archSet := channelArches(v.Platforms)
```

Comment:
Minor: you can have channelArches detect `ArchAll` and return `charm.AllArches`. Then you don't really need the second contains, right?

Reply:
Yes, yes you can.